### PR TITLE
core#1623: My Case dashlet doesn't sort by name but contact_id instead

### PR DESCRIPTION
**Pull request by monishdeb**
Overview
----------------------------------------
My Case dashlet doesn't sort by name but contact_id instead

Steps to replicate:
1. Enable CiviCase component
2. Create multiple case for different contact
3. Go to dashboard and add 'My Cases' dashlet.
4. Try to sort by 'Contact'

Before
----------------------------------------
after

After
----------------------------------------
_Expected behaviour_ : As you can sort by 'Contact' doesn't consider the name. Internally it is sorted by contact ID which is not correct, which should be by sort name.

after

Technical Details
----------------------------------------
The fix is in core BAO fn and its been called in various places in codebase: 
```
Monishs-MacBook-Pro:civicrm-core monish$ grep -irn "CRM_Case_BAO_Case::getCases(" CRM/
CRM//Contact/BAO/Relationship.php:1459:              $userCases = CRM_Case_BAO_Case::getCases(FALSE);
CRM//Dashlet/Page/MyCases.php:49:    if (CRM_Case_BAO_Case::getCases(FALSE, ['type' => 'any'], $context, TRUE)) {
CRM//Dashlet/Page/AllCases.php:49:    if (CRM_Case_BAO_Case::getCases(TRUE, ['type' => 'any'], 'dashboard', TRUE)) {
CRM//Case/Page/DashBoard.php:71:    $upcoming = CRM_Case_BAO_Case::getCases($allCases, [], 'dashboard', TRUE);
CRM//Case/Page/DashBoard.php:72:    $recent = CRM_Case_BAO_Case::getCases($allCases, ['type' => 'recent'], 'dashboard', TRUE);
CRM//Case/Page/AJAX.php:188:    $cases = CRM_Case_BAO_Case::getCases($allCases, $params);
CRM//Case/Page/Tab.php:54:          $userCases = CRM_Case_BAO_Case::getCases(FALSE, ['type' => 'any']);
CRM//Case/Form/Activity.php:96:      $allCases = CRM_Case_BAO_Case::getCases(TRUE, $params);
CRM//Case/BAO/Case.php:1852:      $filterCases = CRM_Case_BAO_Case::getCases(FALSE);
```
Checked the impact of changing the key from contact_id to sort_name on each case and its safe.

### CODE REVIEW

**Comment by monishdeb** [comment]
ping @demeritcowboy @lcdservices @eileenmcnaughton

**Comment by demeritcowboy** [comment]
Thanks @monishdeb looks good. I see some unrelated errors for other columns but it's separate.

* General standards
    * [`r-explain`] __PASS__
    * [`r-user`] __PASS__
    * [`r-doc`] __PASS__
    * [`r-run`] __PASS__
      * When I click on status or type I get a datatables error but I see it on dmaster.demo and local too so is separate.
      * Checked the various places it's used.
* Developer standards
    * [`r-tech`] __PASS__
    * [`r-code`] __PASS__
    * [`r-maint`] __PASS__ Nice test.
    * [`r-test`] __PASS__

**Comment by eileenmcnaughton** [comment]
@monishdeb the actual fix is fine but check the ExportTest for our new (cool) way of handling civiexit in unit tests - we aren't doing stuff like 

```  if ($params['is_unittest']) {
      return $casesDT;
    } ```

anymore

see innCiviExit
```
    if (CIVICRM_UF === 'UnitTests') {
      throw new CRM_Core_Exception_PrematureExitException('civiExit called', $testParameters);
    }
```

**Comment by monishdeb** [comment]
Thanks @eileenmcnaughton for pointing that out. It helped me to avoid tweaking the core AJAX fn. But to still capture the output I need to bring some modification. I have assigned the output as an error data of ```CRM_Core_Exception_PrematureExitException``` so that it can be later used by UTs to compare with the expected result. Please have look now.

File: tests/phpunit/CRM/Case/BAO/CaseTest.php [hunk 1]
```diff
@@ -136,6 +136,113 @@ private function assertCasesOfUser($loggedInUser, $caseId, $caseCount) {
     $this->assertEquals($caseCount, count($caseRoles), 'Total case roles for logged in users must be ' . $caseCount);
   }
 
+  /**
+   * core/issue-1623: My Case dashlet doesn't sort by name but contact_id instead
```
Comment (eileenmcnaughton) [inline tests/phpunit/CRM/Case/BAO/CaseTest.php]: Do you use an IDE -  phpstorm highlights the need to declare the exception here -    * @throws \CRM_Core_Exception

**Comment by eileenmcnaughton** [comment]
@monishdeb I think you missed something off your commit - maybe
```
+++ b/CRM/Utils/JSON.php
@@ -27,7 +27,7 @@ class CRM_Utils_JSON {
   public static function output($input) {
     CRM_Utils_System::setHttpHeader('Content-Type', 'application/json');
     echo json_encode($input);
-    CRM_Utils_System::civiExit();
+    CRM_Utils_System::civiExit(0, $input);
   }
```

**Comment by eileenmcnaughton** [comment]
or perhaps 
```
--- a/CRM/Utils/JSON.php
+++ b/CRM/Utils/JSON.php
@@ -25,9 +25,13 @@ class CRM_Utils_JSON {
    * @param mixed $input
    */
   public static function output($input) {
+    if (CIVICRM_UF === 'UnitTests') {
+      throw new CRM_Core_Exception_PrematureExitException('civiExit called', $input);
+    }
+
```

**Comment by monishdeb** [comment]
argh.. It was in my unsaved changes :( Seema like I need to change my IDE to phpstorm from [Atom](https://ide.atom.io/)

**Comment by eileenmcnaughton** [comment]
:-)

**Comment by eileenmcnaughton** [comment]
@monishdeb my first thought was the change you made but I wonder if 

https://github.com/civicrm/civicrm-core/pull/16647#issuecomment-593200359


is better - because  i happens before messing with  headers &  echo

**Comment by eileenmcnaughton** [comment]
lol snap

**Comment by monishdeb** [comment]
Yup I agree, already updated :)

**Comment by eileenmcnaughton** [comment]
@monishdeb something still wrong with the test

**Pull request merged**